File: src/ios/ACPTarget_Cordova.m
/*
 Copyright 2020 Adobe. All rights reserved.
 This file is licensed to you under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License. You may obtain a copy
 of the License at http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software distributed under
 the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 OF ANY KIND, either express or implied. See the License for the specific language
 governing permissions and limitations under the License.
 */

/********* cordova-acptarget.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import <ACPTarget/ACPTarget.h>
#import <ACPTarget/ACPTargetOrder.h>
#import <ACPTarget/ACPTargetParameters.h>
#import <ACPTarget/ACPTargetRequestObject.h>
#import <ACPTarget/ACPTargetProduct.h>
#import <ACPTarget/ACPTargetPrefetchObject.h>
#import <Cordova/CDVPluginResult.h>


@interface ACPTarget_Cordova : CDVPlugin

- (void)clearPrefetchCache:(CDVInvokedUrlCommand*)command;
- (void)extensionVersion:(CDVInvokedUrlCommand*)command;
- (void)getThirdPartyId:(CDVInvokedUrlCommand*)comman;
- (void)getTntId:(CDVInvokedUrlCommand*)command;
- (void)resetExperience:(CDVInvokedUrlCommand*)command;
- (void)setPreviewRestartDeepLink:(CDVInvokedUrlCommand*)command;
- (void)setThirdPartyId:(CDVInvokedUrlCommand*)command;
- (void)retrieveLocationContent:(CDVInvokedUrlCommand*)command;
- (void)locationClicked:(CDVInvokedUrlCommand*)command;
- (void)locationsDisplayed:(CDVInvokedUrlCommand*)command;
- (void)prefetchContent:(CDVInvokedUrlCommand*)command;


@end

@implementation ACPTarget_Cordova

- (void)clearPrefetchCache:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        [ACPTarget clearPrefetchCache];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)extensionVersion:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        NSString* extensionVersion = [ACPTarget extensionVersion];

        if (extensionVersion != nil && [extensionVersion length] > 0) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:extensionVersion];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}


- (void) getThirdPartyId:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [ACPTarget getThirdPartyId:^(NSString * _Nullable thirdPartyId) {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:thirdPartyId];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }];
}

- (void) getTntId:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [ACPTarget getTntId:^(NSString * _Nullable tntId) {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:tntId];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }];
}

- (void) resetExperience:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        [ACPTarget resetExperience];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) setThirdPartyId:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSString *thirdPartyId = [self getCommandArg:command.arguments[0]];

         [ACPTarget setThirdPartyId:thirdPartyId];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) setPreviewRestartDeepLink:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSURL *deepLink = [self getCommandArg:command.arguments[0]];

		[ACPTarget setPreviewRestartDeeplink:deepLink];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) retrieveLocationContent:(CDVInvokedUrlCommand*)command {
  [self.commandDelegate runInBackground:^{
    NSError *error;
    NSDictionary *locationRequests = (NSDictionary *)command.arguments[0];
    NSMutableArray *requestArray = [NSMutableArray array];
    NSDictionary *profileParam;
    NSDictionary *mboxParam;
    NSDictionary *orderParam;
    NSDictionary *productParam;
    NSDictionary *profileLocParam;
    NSDictionary *mboxLocParam;
    NSDictionary *orderLocParam;
    NSDictionary *productLocParam;
    ACPTargetOrder *orderParameters = nil;
    ACPTargetParameters *targetParameters = nil;
    ACPTargetProduct *productParameter = nil;
    ACPTargetOrder *orderLocParameters = nil;
    ACPTargetParameters *targetLocParameters = nil;
    ACPTargetProduct *productLocParameter = nil;
    for(id key in locationRequests){
        NSDictionary *request =[locationRequests objectForKey:key];
        NSString *mboxName = [request objectForKey:@"mboxName"];
        if([[request objectForKey:@"mboxParameter"] count] > 0){
            mboxParam = [request objectForKey:@"mboxParameter"];            
        }
        if([[request objectForKey:@"profileParameter"] count] > 0){
            profileParam = [request objectForKey:@"profileParameter"];         
        }
        
        if([[request objectForKey:@"orderParameter"] count] > 0){
            orderParam = [request objectForKey:@"orderParameter"];
            if([orderParam objectForKey:@"orderId"] != nil){
                NSString *orderId = [orderParam objectForKey:@"orderId"];
                NSNumber *orderTotal = [orderParam objectForKey:@"orderTotal"];
                NSArray *orderPurchasedIds = [orderParam objectForKey:@"orderPurchasedIds"];
                orderParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
            }
          
        }

        if([[request objectForKey:@"productParameter"] count] > 0){
            productParam = [request objectForKey:@"productParameter"];
            if([productParam objectForKey:@"id"]  != nil){
              NSString *id = [productParam objectForKey:@"id"];
              NSString *categoryIdValue = [productParam objectForKey:@"categoryId"];
              productParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
            }
        }


        targetParameters = [ACPTargetParameters targetParametersWithParameters:mboxParam
            profileParameters:profileParam
            product:productParameter
            order:orderParameters];
            
        ACPTargetRequestObject *requestObject = [ACPTargetRequestObject targetRequestObjectWithName:mboxName targetParameters:targetParameters
            defaultContent:@"defaultContent" callback:^(NSString * _Nullable content) {
                CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
            
        [requestArray addObject:requestObject];
        
    }
    
    
    NSDictionary *requestLoc = (NSDictionary *)command.arguments[1];
    
    if([[requestLoc objectForKey:@"mboxParameter"] count] > 0){
        mboxLocParam = [requestLoc objectForKey:@"mboxParameter"];
        
    }
    if([[requestLoc objectForKey:@"profileParameter"] count] > 0){
        profileLocParam = [requestLoc objectForKey:@"profileParameter"];
        
    }
  
    if([[requestLoc objectForKey:@"orderParameter"] count] > 0){
        orderLocParam = [requestLoc objectForKey:@"orderParameter"];
        if([orderLocParam objectForKey:@"orderId"]  != nil){
            NSString *orderId = [orderLocParam objectForKey:@"orderId"];
            NSNumber *orderTotal = [orderLocParam objectForKey:@"orderTotal"];
            NSArray *orderPurchasedIds = [orderLocParam objectForKey:@"orderPurchasedIds"];
            orderLocParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
        }
       
    }

    if([[requestLoc objectForKey:@"productParameter"] count] > 0){
        productLocParam = [requestLoc objectForKey:@"productParameter"];
        if([productLocParam objectForKey:@"id"]  != nil){
          NSString *id = [productLocParam objectForKey:@"id"];
          NSString *categoryIdValue = [productLocParam objectForKey:@"categoryId"];
          productLocParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
        }
        
    }
    
    targetLocParameters = [ACPTargetParameters targetParametersWithParameters:mboxLocParam
                                                    profileParameters:profileLocParam
                                                              product:productLocParameter
                                                                order:orderLocParameters];
                                                                
    [ACPTarget retrieveLocationContent:requestArray withParameters:targetLocParameters];

  }];
}

- (void) locationClicked:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
    
        NSString *mboxName = [self getCommandArg:command.arguments[0]];
        NSDictionary *request = [self getCommandArg:command.arguments[1]];
        NSDictionary *profileParam;
        NSDictionary *mboxParam;
        NSDictionary *orderParam;
        NSDictionary *productParam;
        ACPTargetOrder *orderParameters = nil;
        ACPTargetParameters *targetParameters = nil;
        ACPTargetProduct *productParameter = nil;
        
      
        if([[request objectForKey:@"mboxParameter"] count] > 0){
            mboxParam = [request objectForKey:@"mboxParameter"];
        }
        if([[request objectForKey:@"profileParameter"] count] > 0){
            profileParam = [request objectForKey:@"profileParameter"];
        }
      
        if([[request objectForKey:@"orderParameter"] count] > 0){
            orderParam = [request objectForKey:@"orderParameter"];
            if([orderParam objectForKey:@"orderId"] != nil){
                NSString *orderId = [orderParam objectForKey:@"orderId"];
                NSNumber *orderTotal = [orderParam objectForKey:@"orderTotal"];
                NSArray *orderPurchasedIds = [orderParam objectForKey:@"orderPurchasedIds"];
                orderParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
            }
        }
        
        if([[request objectForKey:@"productParameter"] count] > 0){
            productParam = [request objectForKey:@"productParameter"];
            if([productParam objectForKey:@"id"]  != nil){
              NSString *id = [productParam objectForKey:@"id"];
              NSString *categoryIdValue = [productParam objectForKey:@"categoryId"];
              productParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
            }
        }
        
        targetParameters = [ACPTargetParameters targetParametersWithParameters:mboxParam
            profileParameters:profileParam
            product:productParameter
            order:orderParameters];
                                                                
        [ACPTarget locationClickedWithName:mboxName targetParameters:targetParameters];
        
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) locationsDisplayed:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSArray *mboxLists = [self getCommandArg:command.arguments[0]];
        NSDictionary *request = [self getCommandArg:command.arguments[1]];
        NSDictionary *profileParam;
        NSDictionary *mboxParam;
        NSDictionary *orderParam;
        NSDictionary *productParam;
        ACPTargetOrder *orderParameters = nil;
        ACPTargetParameters *targetParameters = nil;
        ACPTargetProduct *productParameter = nil;
        

        
        if([[request objectForKey:@"mboxParameter"] count] > 0){
            mboxParam = [request objectForKey:@"mboxParameter"];
        }
        if([[request objectForKey:@"profileParameter"] count] > 0){
            profileParam = [request objectForKey:@"profileParameter"];
        }
      
        if([[request objectForKey:@"orderParameter"] count] > 0){
            orderParam = [request objectForKey:@"orderParameter"];
            if([orderParam objectForKey:@"orderId"] != nil){
                NSString *orderId = [orderParam objectForKey:@"orderId"];
                NSNumber *orderTotal = [orderParam objectForKey:@"orderTotal"];
                NSArray *orderPurchasedIds = [orderParam objectForKey:@"orderPurchasedIds"];
                orderParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
            }
        }
        
        if([[request objectForKey:@"productParameter"] count] > 0){
            productParam = [request objectForKey:@"productParameter"];
            if([productParam objectForKey:@"id"]  != nil){
              NSString *id = [productParam objectForKey:@"id"];
              NSString *categoryIdValue = [productParam objectForKey:@"categoryId"];
              productParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
            }
        }
        
        targetParameters = [ACPTargetParameters targetParametersWithParameters:mboxParam
            profileParameters:profileParam
            product:productParameter
            order:orderParameters];
        
        
        [ACPTarget locationsDisplayed:mboxLists withTargetParameters:targetParameters];
        
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) prefetchContent:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        NSError *error;
        NSDictionary *locationRequests = (NSDictionary *)command.arguments[0];
        NSMutableArray *requestArray = [NSMutableArray array];
        NSDictionary *profileParam;
        NSDictionary *mboxParam;
        NSDictionary *orderParam;
        NSDictionary *productParam;
        NSDictionary *profileLocParam;
        NSDictionary *mboxLocParam;
        NSDictionary *orderLocParam;
        NSDictionary *productLocParam;
        ACPTargetOrder *orderParameters = nil;
        ACPTargetParameters *targetParameters = nil;
        ACPTargetProduct *productParameter = nil;
        ACPTargetOrder *orderLocParameters = nil;
        ACPTargetParameters *targetLocParameters = nil;
        ACPTargetProduct *productLocParameter = nil;
        for(id key in locationRequests){
            NSDictionary *request =[locationRequests objectForKey:key];
            NSString *mboxName = [request objectForKey:@"mboxName"];
            if([[request objectForKey:@"mboxParameter"] count] > 0){
                mboxParam = [request objectForKey:@"mboxParameter"];
            }
            if([[request objectForKey:@"profileParameter"] count] > 0){
                profileParam = [request objectForKey:@"profileParameter"];
            }
          
            if([[request objectForKey:@"orderParameter"] count] > 0){
                orderParam = [request objectForKey:@"orderParameter"];
                if([orderParam objectForKey:@"orderId"] != nil){
                    NSString *orderId = [orderParam objectForKey:@"orderId"];
                    NSNumber *orderTotal = [orderParam objectForKey:@"orderTotal"];
                    NSArray *orderPurchasedIds = [orderParam objectForKey:@"orderPurchasedIds"];
                    orderParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
                }
            }

            if([[request objectForKey:@"productParameter"] count] > 0){
                productParam = [request objectForKey:@"productParameter"];
                if([productParam objectForKey:@"id"]  != nil){
                  NSString *id = [productParam objectForKey:@"id"];
                  NSString *categoryIdValue = [productParam objectForKey:@"categoryId"];
                  productParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
                }
            }


            targetParameters = [ACPTargetParameters targetParametersWithParameters:mboxParam
                profileParameters:profileParam
                product:productParameter
                order:orderParameters];
            
            ACPTargetPrefetchObject *requestObject = [ACPTargetPrefetchObject targetPrefetchObjectWithName:mboxName
                targetParameters:targetParameters];
                
            [requestArray addObject:requestObject];
        }
        
        
        NSDictionary *requestLoc = (NSDictionary *)command.arguments[1];
        
        
        if([[requestLoc objectForKey:@"mboxParameter"] count] > 0){
            mboxLocParam = [requestLoc objectForKey:@"mboxParameter"];
        }
        if([[requestLoc objectForKey:@"profileParameter"] count] > 0){
            profileLocParam = [requestLoc objectForKey:@"profileParameter"];
        }
      
        if([[requestLoc objectForKey:@"orderParameter"] count] > 0){
            orderLocParam = [requestLoc objectForKey:@"orderParameter"];
            if([orderLocParam objectForKey:@"orderId"]  != nil){
                NSString *orderId = [orderLocParam objectForKey:@"orderId"];
                NSNumber *orderTotal = [orderLocParam objectForKey:@"orderTotal"];
                NSArray *orderPurchasedIds = [orderLocParam objectForKey:@"orderPurchasedIds"];
                orderLocParameters = [ACPTargetOrder targetOrderWithId:orderId total:orderTotal purchasedProductIds:orderPurchasedIds];
            }
        }

        if([[requestLoc objectForKey:@"productParameter"] count] > 0){
            productLocParam = [requestLoc objectForKey:@"productParameter"];
            if([productLocParam objectForKey:@"id"]  != nil){
              NSString *id = [productLocParam objectForKey:@"id"];
              NSString *categoryIdValue = [productLocParam objectForKey:@"categoryId"];
              productLocParameter = [ACPTargetProduct targetProductWithId:id categoryId:categoryIdValue];
            }
        }
        
        targetLocParameters = [ACPTargetParameters targetParametersWithParameters:mboxLocParam
                                                        profileParameters:profileLocParam
                                                                  product:productLocParameter
                                                                    order:orderLocParameters];
                                                                    
        [ACPTarget prefetchContent:requestArray withParameters:targetLocParameters callback:^(NSError * _Nullable error) {
                CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
        
        

      }];
}

/*
 * Helper functions
 */

- (id) getCommandArg:(id) argument {
    return argument == (id)[NSNull null] ? nil : argument;
}

@end

File: src/ios/ACPTarget_Cordova.swift
import AEPTarget

@objc(ACPTarget_Cordova) class ACPTarget_Cordova: CDVPlugin {

  @objc(clearPrefetchCache:)
  func clearPrefetchCache(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      Target.clearPrefetchCache()
      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(extensionVersion:)
  func extensionVersion(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      var pluginResult: CDVPluginResult! = nil
      let extensionVersion: String! = Target.extensionVersion

      if extensionVersion != nil && extensionVersion.count > 0 {
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: extensionVersion)
      } else {
        pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)
      }

      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(getThirdPartyId:)
  func getThirdPartyId(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      Target.getThirdPartyId({ (thirdPartyId: String?, error) in
        if error == nil {
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_OK, messageAs: thirdPartyId)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_ERROR, messageAs: error?.localizedDescription)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
      })
    })
  }

  @objc(getTntId:)
  func getTntId(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      Target.getTntId({ (tntId: String?, error) in

        if error == nil {
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_OK, messageAs: tntId)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
          let pluginResult: CDVPluginResult! = CDVPluginResult(
            status: CDVCommandStatus_ERROR, messageAs: error?.localizedDescription)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }

      })
    })
  }

  @objc(resetExperience:)
  func resetExperience(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      Target.resetExperience()
      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(setThirdPartyId:)
  func setThirdPartyId(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      let thirdPartyId: String! = command.arguments[0] as? String

      Target.setThirdPartyId(thirdPartyId)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(setPreviewRestartDeepLink:)
  func setPreviewRestartDeepLink(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      guard let deepLink = URL(string: command.arguments[0] as! String) else {
        self.commandDelegate.send(
          CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Unable setPreviewRestartDeepLink. Input was malformed"),
          callbackId: command.callbackId)
        return
      }

      Target.setPreviewRestartDeepLink(deepLink)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(retrieveLocationContent:)
func retrieveLocationContent(command: CDVInvokedUrlCommand) {
    self.commandDelegate.run(inBackground: {
        guard let locationRequests = command.arguments[0] as? [String: [String: Any]] else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid location requests")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        var requestArray: [TargetRequest] = []
        for (_, requestDict) in locationRequests {
            guard let mboxName = requestDict["mboxName"] as? String else { continue }
            let mboxParam = requestDict["mboxParameter"] as? [String: String]
            let profileParam = requestDict["profileParameter"] as? [String: String]
            var orderParameters: TargetOrder?
            if let orderParam = requestDict["orderParameter"] as? [String: Any],
               let orderId = orderParam["orderId"] as? String,
               let orderTotal = orderParam["orderTotal"] as? Double,
               let purchasedIds = orderParam["orderPurchasedIds"] as? [String] {
                orderParameters = TargetOrder(id: orderId, total: orderTotal, purchasedProductIds: purchasedIds)
            }
            var productParameter: TargetProduct?
            if let productParam = requestDict["productParameter"] as? [String: Any],
               let id = productParam["id"] as? String,
               let categoryId = productParam["categoryId"] as? String {
                productParameter = TargetProduct(productId: id, categoryId: categoryId)
            }
            let targetParameters = TargetParameters(
                parameters: mboxParam,
                profileParameters: profileParam,
                order: orderParameters,
                product: productParameter)
            let requestObject = TargetRequest(mboxName: mboxName, defaultContent: "defaultContent", targetParameters: targetParameters, contentCallback: nil)
            requestArray.append(requestObject)
        }
        guard let requestLoc = command.arguments[1] as? [String: Any] else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid request location")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        let mboxLocParam = requestLoc["mboxParameter"] as? [String: String]
        let profileLocParam = requestLoc["profileParameter"] as? [String: String]
        var orderLocParameters: TargetOrder?
        if let orderLocParam = requestLoc["orderParameter"] as? [String: Any],
           let orderId = orderLocParam["orderId"] as? String,
           let orderTotal = orderLocParam["orderTotal"] as? Double,
           let purchasedIds = orderLocParam["orderPurchasedIds"] as? [String] {
            orderLocParameters = TargetOrder(id: orderId, total: orderTotal, purchasedProductIds: purchasedIds)
        }
        var productLocParameter: TargetProduct?
        if let productLocParam = requestLoc["productParameter"] as? [String: Any],
           let id = productLocParam["id"] as? String,
           let categoryId = productLocParam["categoryId"] as? String {
            productLocParameter = TargetProduct(productId: id, categoryId: categoryId)
        }
        let targetLocParameters = TargetParameters(
            parameters: mboxLocParam,
            profileParameters: profileLocParam,
            order: orderLocParameters,
            product: productLocParameter)
        Target.retrieveLocationContent(requestArray, with: targetLocParameters)
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
}

  @objc(locationClicked:)
  func locationClicked(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let mboxName: String! = command.arguments[0] as? String
      let request: NSDictionary! = command.arguments[1] as? NSDictionary
      var profileParam: NSDictionary!
      var mboxParam: NSDictionary!
      var orderParam: NSDictionary!
      var productParam: NSDictionary!
      var orderParameters: TargetOrder! = nil
      var targetParameters: TargetParameters! = nil
      var productParameter: TargetProduct! = nil

      if request.object(forKey: "mboxParameter") != nil {
        mboxParam = request.object(forKey: "mboxParameter") as? NSDictionary
      }
      if request.object(forKey: "profileParameter") != nil {
        profileParam = request.object(forKey: "profileParameter") as? NSDictionary
      }

      if request.object(forKey: "orderParameter") != nil {
        orderParam = request.object(forKey: "orderParameter") as? NSDictionary
        if orderParam.object(forKey: "orderId") != nil {
          let orderId: String! = orderParam.object(forKey: "orderId") as? String
          let orderTotal: Double! = orderParam.object(forKey: "orderTotal") as? Double
          let orderPurchasedIds: [String]! =
            orderParam.object(forKey: "orderPurchasedIds") as? [String]
          orderParameters = TargetOrder(
            id: orderId, total: orderTotal, purchasedProductIds: orderPurchasedIds)
        }
      }

      if request.object(forKey: "productParameter") != nil {
        productParam = request.object(forKey: "productParameter") as? NSDictionary
        if productParam.object(forKey: "id") != nil {
          let id: String! = productParam.object(forKey: "id") as? String
          let categoryIdValue: String! = productParam.object(forKey: "categoryId") as? String
          productParameter = TargetProduct(productId: id, categoryId: categoryIdValue)
        }
      }

      targetParameters = TargetParameters(
        parameters: mboxParam as? [String: String],
        profileParameters: profileParam as? [String: String],
        order: orderParameters, product: productParameter)

      Target.clickedLocation(mboxName, targetParameters: targetParameters)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(locationsDisplayed:)
  func locationsDisplayed(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {
      let mboxLists: [String]! = command.arguments[0] as? [String]
      let request: NSDictionary! = command.arguments[1] as? NSDictionary
      var profileParam: NSDictionary!
      var mboxParam: NSDictionary!
      var orderParam: NSDictionary!
      var productParam: NSDictionary!
      var orderParameters: TargetOrder! = nil
      var targetParameters: TargetParameters! = nil
      var productParameter: TargetProduct! = nil

      if request.object(forKey: "mboxParameter") != nil {
        mboxParam = request.object(forKey: "mboxParameter") as? NSDictionary
      }
      if request.object(forKey: "profileParameter") != nil {
        profileParam = request.object(forKey: "profileParameter") as? NSDictionary
      }

      if request.object(forKey: "orderParameter") != nil {
        orderParam = request.object(forKey: "orderParameter") as? NSDictionary
        if orderParam.object(forKey: "orderId") != nil {
          let orderId: String! = orderParam.object(forKey: "orderId") as? String
          let orderTotal: Double! = orderParam.object(forKey: "orderTotal") as? Double
          let orderPurchasedIds: [String]! =
            orderParam.object(forKey: "orderPurchasedIds") as? [String]
          orderParameters = TargetOrder(
            id: orderId, total: orderTotal, purchasedProductIds: orderPurchasedIds)
        }
      }

      if request.object(forKey: "productParameter") != nil {
        productParam = request.object(forKey: "productParameter") as? NSDictionary
        if productParam.object(forKey: "id") != nil {
          let id: String! = productParam.object(forKey: "id") as? String
          let categoryIdValue: String! = productParam.object(forKey: "categoryId") as? String
          productParameter = TargetProduct(productId: id, categoryId: categoryIdValue)
        }
      }

      targetParameters = TargetParameters(
        parameters: mboxParam as? [String: String],
        profileParameters: profileParam as? [String: String],
        order: orderParameters, product: productParameter)

      Target.displayedLocations(mboxLists, targetParameters: targetParameters)

      let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
      self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    })
  }

  @objc(prefetchContent:)
  func prefetchContent(command: CDVInvokedUrlCommand!) {
    self.commandDelegate.run(inBackground: {

      let locationRequests: NSDictionary! = command.arguments[0] as? NSDictionary
      let requestArray: NSMutableArray! = NSMutableArray.init()
      var profileParam: NSDictionary!
      var mboxParam: NSDictionary!
      var orderParam: NSDictionary!
      var productParam: NSDictionary!
      var profileLocParam: NSDictionary!
      var mboxLocParam: NSDictionary!
      var orderLocParam: NSDictionary!
      var productLocParam: NSDictionary!
      var orderParameters: TargetOrder! = nil
      var targetParameters: TargetParameters! = nil
      var productParameter: TargetProduct! = nil
      var orderLocParameters: TargetOrder! = nil
      var targetLocParameters: TargetParameters! = nil
      var productLocParameter: TargetProduct! = nil

      locationRequests.forEach { (key: Any, valueObj: Any) in
        let request: NSDictionary! = valueObj as? NSDictionary
        let mboxName: String! = request.object(forKey: "mboxName") as? String

        if request.object(forKey: "mboxParameter") != nil {
          mboxParam = request.object(forKey: "mboxParameter") as? NSDictionary
        }
        if request.object(forKey: "profileParameter") != nil {
          profileParam = request.object(forKey: "profileParameter") as? NSDictionary
        }

        if request.object(forKey: "orderParameter") != nil {
          orderParam = request.object(forKey: "orderParameter") as? NSDictionary
          if orderParam.object(forKey: "orderId") != nil {
            let orderId: String! = orderParam.object(forKey: "orderId") as? String
            let orderTotal: Double! = orderParam.object(forKey: "orderTotal") as? Double
            let orderPurchasedIds: [String]! =
              orderParam.object(forKey: "orderPurchasedIds") as? [String]
            orderParameters = TargetOrder(
              id: orderId, total: orderTotal, purchasedProductIds: orderPurchasedIds)
          }
        }

        if request.object(forKey: "productParameter") != nil {
          productParam = request.object(forKey: "productParameter") as? NSDictionary
          if productParam.object(forKey: "id") != nil {
            let id: String! = productParam.object(forKey: "id") as? String
            let categoryIdValue: String! = productParam.object(forKey: "categoryId") as? String
            productParameter = TargetProduct(productId: id, categoryId: categoryIdValue)
          }
        }

        targetParameters = TargetParameters(
          parameters: mboxParam as? [String: String],
          profileParameters: profileParam as? [String: String],
          order: orderParameters, product: productParameter)

        let requestObject: TargetPrefetch! = TargetPrefetch(
          name: mboxName,
          targetParameters: targetParameters)

        requestArray.add(requestObject as Any)

      }

      let requestLoc: NSDictionary! = command.arguments[1] as? NSDictionary

      if requestLoc.object(forKey: "mboxParameter") != nil {
        mboxLocParam = requestLoc.object(forKey: "mboxParameter") as? NSDictionary
      }
      if requestLoc.object(forKey: "profileParameter") != nil {
        profileLocParam = requestLoc.object(forKey: "profileParameter") as? NSDictionary
      }

      if requestLoc.object(forKey: "orderParameter") != nil {
        orderLocParam = requestLoc.object(forKey: "orderParameter") as? NSDictionary
        if orderLocParam.object(forKey: "orderId") != nil {
          let orderId: String! = orderLocParam.object(forKey: "orderId") as? String
          let orderTotal: Double! = orderLocParam.object(forKey: "orderTotal") as? Double
          let orderPurchasedIds: [String]! =
            orderLocParam.object(forKey: "orderPurchasedIds") as? [String]
          orderLocParameters = TargetOrder(
            id: orderId, total: orderTotal, purchasedProductIds: orderPurchasedIds)
        }
      }

      if requestLoc.object(forKey: "productParameter") != nil {
        productLocParam = requestLoc.object(forKey: "productParameter") as? NSDictionary
        if productLocParam.object(forKey: "id") != nil {
          let id: String! = productLocParam.object(forKey: "id") as? String
          let categoryIdValue: String! = productLocParam.object(forKey: "categoryId") as? String
          productLocParameter = TargetProduct(productId: id, categoryId: categoryIdValue)
        }
      }

      targetLocParameters = TargetParameters(
        parameters: mboxLocParam as? [String: String],
        profileParameters: profileLocParam as? [String: String],
        order: orderLocParameters, product: productLocParameter)

      Target.prefetchContent(
        requestArray as! [TargetPrefetch], with: targetLocParameters,
        { (error) in
          let pluginResult: CDVPluginResult! = CDVPluginResult(status: CDVCommandStatus_OK)
          self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })

    })
  }
}
